core/src/main/java/io/grpc/Context.java

```diff
@@ -0,0 +1,653 @@
+   */
+  public void attach() {
+    contextStack.get().addLast(this);
+  }
+
+  /**
+   * Returns {@code true} if this context is {@link #current}, {@code false} otherwise.
+   */
+  public boolean isCurrent() {
+    return current() == this;
+  }
+
+  /**
+   * Detach the current context from the thread and restore the context that was previously
+   * attached to the thread as the 'current' context.
+   *
+   * @throws java.lang.IllegalStateException if this context is not {@link #current()}.
+   */
+  public void detach() {
+    ArrayDeque<Context> stack = contextStack.get();
+    if (stack.isEmpty()) {
+      if (this == ROOT) {
+        throw new IllegalStateException("Cannot detach root");
+      } else {
+        throw new IllegalStateException("Cannot detach non-root context when root is current");
+      }
+    }
+    if (stack.peekLast() != this) {
+      throw new IllegalStateException("Cannot detach a context that is not current");
+    }
+    stack.removeLast();
+  }
+
+  /**
+   * Is this context cancelled.
+   */
+  public boolean isCancelled() {
+    if (parent == null || !cascadesCancellation) {
+      return false;
+    } else {
+      return parent.isCancelled();
+    }
+  }
+
+  /**
+   * Return the cause of the cancellation or {@code null} if context was cancelled without a cause.
+   */
+  @Nullable
+  public Throwable cause() {
+    if (parent == null || !cascadesCancellation) {
+      return null;
+    } else {
+      return parent.cause();
+    }
+  }
+
+  /**
+   * Add a listener that will be notified when the context becomes cancelled.
+   */
+  public void addListener(final CancellationListener cancellationListener,
+                          final Executor executor) {
+    Preconditions.checkNotNull(cancellationListener);
+    Preconditions.checkNotNull(executor);
+    if (canBeCancelled()) {
+      ExecutableListener executableListener =
+          new ExecutableListener(executor, cancellationListener);
+      synchronized (this) {
+        if (isCancelled()) {
+          executableListener.deliver();
+        } else {
+          if (listeners == null) {
+            // Now that we have a listener we need to listen to our parent so
+            // we can cascade listener notification.
+            listeners = new ArrayList<ExecutableListener>();
+            if (parent.canBeCancelled()) {
+              parent.addListener(parentListener, MoreExecutors.directExecutor());
+            }
+          }
+          listeners.add(executableListener);
+        }
+      }
+    } else {
+      // Discussion point: Should we throw or suppress.
+    }
+  }
+
+  /**
+   * Remove a {@link CancellationListener}.
+   */
+  public void removeListener(CancellationListener cancellationListener) {
+    synchronized (this) {
+      if (listeners != null) {
+        for (int i = listeners.size() - 1; i >= 0; i--) {
+          if (listeners.get(i).listener == cancellationListener) {
+            listeners.remove(i);
```

Comment:
Right, _because_ we aren't suppressing duplicate listeners, it seems remove should only remove one listener.

```java
context.addListener(this);
context.addListener(this);
// Here `this` gets notified twice
context.removeListener(this);
// But here it doesn't get notified at all
```

If we coalesced duplicate adds (like a set does) then the current remove behavior makes sense. But given we are treating this as a simple collection of listeners where duplicates are allowed, removal removing all instances is unexpected.